src/main/java/core/PyTAG.java

```diff
@@ -367,86 +380,298 @@ public int sampleRNDAction(int[] mask, Random rnd){
+                }
 
-                    // step the environment
-                    env.step(randomAction);
-                } else {
-                    // this is the normal game loop
-                    List<AbstractAction> actions = env.forwardModel.computeAvailableActions(env.gameState);
+                int[] maskBefore = env.getActionMask();
+                int legalBefore = 0; for (int v : maskBefore) if (v != 0) legalBefore++;
+                System.out.printf("Phase: %s | turnOwner: %d | currentActor: %d | inProgress: %s%n",
+                        env.getGamePhase(), env.getTurnOwner(), env.getCurrentActor(), env.getActionInProgressName());
+                System.out.println("BEFORE legal actions: " + legalBefore);
+                for (int i = 0; i < maskBefore.length; i++) {
+                    if (maskBefore[i] == 1) {
+                        System.out.printf("  [%d] %s%n", i, env.getActionNameById(i));
+                    }
+                }
+
+                // choose a random valid action
+                int[] validIdx = java.util.stream.IntStream.range(0, maskBefore.length)
+                        .filter(i -> maskBefore[i] == 1).toArray();
+                if (validIdx.length == 0) {
+                    System.out.println("⚠️  No valid actions; stopping.");
+                    break;
+                }
+                int actionId = validIdx[rnd.nextInt(validIdx.length)];
+                System.out.println("Chosen actionId: " + actionId + " -> " + env.getActionNameById(actionId));
+
+                // Step
+                env.step(actionId);
+
+                // --- After step ---
+                System.out.printf("%n=== AFTER step %d ===%n", step);
+                System.out.println("Phase: " + env.getGamePhase()
+                        + " | turnOwner: " + env.getTurnOwner()
+                        + " | currentActor: " + env.getCurrentActor()
+                        + " | inProgress: " + env.getActionInProgressName());
+                int[] maskAfter = env.getActionMask();
+                int legalAfter = 0; for (int v : maskAfter) if (v != 0) legalAfter++;
+                System.out.println("AFTER legal actions: " + legalAfter);
+                for (int i = 0; i < maskAfter.length; i++) {
+                    if (maskAfter[i] == 1) {
+                        System.out.printf("  [%d] %s%n", i, env.getActionNameById(i));
+                    }
+                }
+            }
+        } catch (Exception e) {
+            System.out.println("Exception during game initialisation: " + e);
+            e.printStackTrace();
+        }
+    }
 
-                    double reward = env.getReward();
-//                    System.out.println("at step " + steps + " the reward is " + reward + "player ID " + env.gameState.getCurrentPlayer());
+    
+ // --- ADDED FOR DEBUG DELETE LATER ---
```

Comment:
PyTAG.java should not be left with any game-specific code in it (and I think this can all be safely removed). It would also be better to leave the example game in main() as TicTacToe as a better starting example.